### PR TITLE
extent_delegate_list

### PR DESCRIPTION
**Pull request by jacob314**
This class forks and extends the RenderSliverFixedExtentBoxAdaptor from package:flutter to support
lists where each element can have a custom extent provided by a delegate.
This makes animating the inspector tree practical and will greatly improve performance in the flame chart.
This list shares the benefit of the ListView constructors specifying a fixed extent in that only list view items in view need to built instead of building all items to measure the list.
It still makes sense to leave RenderSliverFixedExtentBoxAdaptor as is in package:flutter as completely fixed extents are a bit more efficient than this implementation. For example, there are const object optimizations only practical in the fixed case and a typical implementation of ExtentDelegate will take O(log n) for some opterations that are O(1) for the fixed case.

### CODE REVIEW

File: packages/devtools_app/lib/src/flutter/extent_delegate_list.dart [hunk 1]
```diff
@@ -0,0 +1,459 @@
+import 'dart:math' as math;
+
+import 'package:collection/collection.dart' as collection;
+import 'package:flutter/foundation.dart';
+import 'package:flutter/rendering.dart';
+import 'package:flutter/widgets.dart';
+
+/// Base class for delegate providing extent information for items in a list.
+abstract class ExtentDelegate {
```
Comment (jacob314) [inline packages/devtools_app/lib/src/flutter/extent_delegate_list.dart]: @kenzieschmoll I expect you will want to provide your own implementation of this class to leverage the fact that you know the absolute position of each element up front rather than really wanting to reason in terms of adding up sizes of each element. You might want to move some of the code in FixedExtentDelegate to a shared based class that handles binary searching for the matching row. The other difference is you probably ideally want to support gaps between rows. There isn't any reason this class couldn't support that although you will need to tweak the layout method as it currently assumes zero gaps so some of the logic is a bit odd if you have gaps.

File: packages/devtools_app/lib/src/flutter/extent_delegate_list.dart [hunk 2]
```diff
@@ -0,0 +1,459 @@
...
+
+  final double Function(int index) computeExtent;
+  final int Function() computeLength;
+
+  @override
+  void recompute() {
+    final length = computeLength();
+    _totalExtent = List(length + 1);
+    double total = 0;
+    _totalExtent[0] = 0;
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/flutter/extent_delegate_list.dart]: Why are we setting this?

File: packages/devtools_app/lib/src/flutter/extent_delegate_list.dart [hunk 3]
```diff
@@ -0,0 +1,459 @@
...
+
+/// Base class for delegate providing extent information for items in a list.
+abstract class ExtentDelegate {
+  /// Optional callback to execute after the layout of the extents is modified.
+  VoidCallback onLayoutDirty;
+
+  int get length;
+
+  /// The main-axis extent of each item.
+  double getItemExtent(int index);
```
[the remark this answers is not in the document]
Comment (jacob314) [reply packages/devtools_app/lib/src/flutter/extent_delegate_list.dart]: these bad method names do match the bad method names used in flutter so I am slightly torn on whether to make this match Dart style.. Switched it to Dart style anyway as Flutter would probably change if it wasn't a breaking change.

File: packages/devtools_app/lib/src/flutter/extent_delegate_list.dart [hunk 4]
```diff
@@ -0,0 +1,459 @@
...
+
+  /// The layout offset for the child with the given index.
+  double indexToLayoutOffset(int index);
```
[the remark this answers is not in the document]
Comment (jacob314) [reply packages/devtools_app/lib/src/flutter/extent_delegate_list.dart]: renamed to just `layoutOffset`.

File: packages/devtools_app/lib/src/flutter/extent_delegate_list.dart [hunk 5]
```diff
@@ -0,0 +1,459 @@
...
+  @override
+  int get length => _totalExtent.length - 1;
+
+  final double Function(int index) computeExtent;
+  final int Function() computeLength;
+
+  @override
+  void recompute() {
+    final length = computeLength();
+    _totalExtent = List(length + 1);
```
[the remark this answers is not in the document]
Comment (jacob314) [reply packages/devtools_app/lib/src/flutter/extent_delegate_list.dart]: documented. Renamed _totalExent to offsets to help clarify. _totalExtend was misleading.

File: packages/devtools_app/lib/src/flutter/extent_delegate_list.dart (hunk 2 above)
Comment (jacob314) [reply packages/devtools_app/lib/src/flutter/extent_delegate_list.dart]: commented. We should remove once we have NNBD

File: packages/devtools_app/lib/src/flutter/extent_delegate_list.dart [hunk 6]
```diff
@@ -0,0 +1,464 @@
...
+      index--;
+    }
+    assert(_offsets[index] <= scrollOffset + precisionErrorTolerance);
+    return index;
+  }
+
+  @override
+  int maxChildIndexForScrollOffset(double scrollOffset) {
+    int index = collection.lowerBound(_offsets, scrollOffset);
+    // TODO(jacobr): review.
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/flutter/extent_delegate_list.dart]: nit: do we still need this todo?

File: packages/devtools_app/lib/src/flutter/extent_delegate_list.dart [hunk 7]
```diff
@@ -0,0 +1,464 @@
...
+          shrinkWrap: shrinkWrap,
+          padding: padding,
+          semanticChildCount: semanticChildCount,
+        );
+
+  /// A delegate that provides the children for the [ExtentDelegateListView].
+  final SliverChildDelegate childrenDelegate;
+
+  /// A delegate that provides item extents for the children of the
+  ///  [ExtentDelegateListView].
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/flutter/extent_delegate_list.dart]: super nit: remove extra space at beginning of line

File: packages/devtools_app/lib/src/flutter/extent_delegate_list.dart [hunk 8]
```diff
@@ -0,0 +1,464 @@
...
+    return SliverExtentDelegateList(
+      delegate: childrenDelegate,
+      extentDelegate: extentDelegate,
+    );
+  }
+}
+
+/// A sliver that places multiple box children in a linear array.
+///
+/// The  main axis extents on each child are specified by a delegate.
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/flutter/extent_delegate_list.dart]: nit: extra space

File: packages/devtools_app/lib/src/flutter/extent_delegate_list.dart [hunk 9]
```diff
@@ -0,0 +1,464 @@
...
+  }) : super(key: key, delegate: delegate);
+
+  /// The extent the children are forced to have in the main axis.
+  final ExtentDelegate extentDelegate;
+
+  @override
+  RenderSliverExtentDelegateBoxAdaptor createRenderObject(
+      BuildContext context) {
+    final SliverMultiBoxAdaptorElement element =
+        context as SliverMultiBoxAdaptorElement;
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/flutter/extent_delegate_list.dart]: do we need both the explicit type and the cast here?

File: packages/devtools_app/lib/src/flutter/extent_delegate_list.dart [hunk 10]
```diff
@@ -0,0 +1,464 @@
...
+
+    if (firstChild != null) {
+      final int leadingGarbage = _calculateLeadingGarbage(firstIndex);
+      final int trailingGarbage = _calculateTrailingGarbage(targetLastIndex);
+      collectGarbage(leadingGarbage, trailingGarbage);
+    } else {
+      collectGarbage(0, 0);
+    }
+
+    if (firstChild == null) {
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/flutter/extent_delegate_list.dart]: can we put all of this in the previous else statement?

File: packages/devtools_app/lib/src/flutter/extent_delegate_list.dart [hunk 11]
```diff
@@ -0,0 +1,464 @@
...
+      if (!addInitialChild(
+          index: firstIndex,
+          layoutOffset: _extentDelegate.layoutOffset(firstIndex))) {
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/flutter/extent_delegate_list.dart]: nit: trailing comma?

File: packages/devtools_app/lib/src/flutter/extent_delegate_list.dart [hunk 12]
```diff
@@ -0,0 +1,464 @@
...
+      minExtent: currentItemExtent,
+      maxExtent: currentItemExtent,
+    );
+  }
+
+  // This method is is a fork of RenderSliverFixedExtentBoxAdaptor.performLayout
+  // where we defer computations about offsets to the _extendDelegate and try
+  // to avoid logic only applicable if all children have the same extent.
+  @override
+  void performLayout() {
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/flutter/extent_delegate_list.dart]: can any of this method be put into helpers to make it more readable? Or are you trying to keep a similar structure to the flutter framework fork?

File: packages/devtools_app/lib/src/flutter/extent_delegate_list.dart [hunk 13]
```diff
@@ -0,0 +1,464 @@
...
+
+    for (int index = indexOf(firstChild) - 1; index >= firstIndex; --index) {
+      final RenderBox child =
+          insertAndLayoutLeadingChild(buildChildConstraints(index));
+      if (child == null) {
+        // Items before the previously first child are no longer present.
+        // Reset the scroll offset to offset all items prior and up to the
+        // missing item. Let parent re-layout everything.
+        geometry = SliverGeometry(
+            scrollOffsetCorrection: _extentDelegate.layoutOffset(index));
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/flutter/extent_delegate_list.dart]: trailing comma

File: packages/devtools_app/lib/src/flutter/extent_delegate_list.dart [hunk 14]
```diff
@@ -0,0 +1,464 @@
...
+        return;
+      }
+      final SliverMultiBoxAdaptorParentData childParentData =
+          child.parentData as SliverMultiBoxAdaptorParentData;
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/flutter/extent_delegate_list.dart]: same question here and elsewhere about whether both the type and the cast are necessary

File: packages/devtools_app/lib/src/flutter/extent_delegate_list.dart [hunk 15]
```diff
@@ -0,0 +1,464 @@
...
+
+    double estimatedMaxScrollOffset =
+        _extentDelegate.layoutOffset(_extentDelegate.length);
+    for (int index = indexOf(trailingChildWithLayout) + 1;
+        targetLastIndex == null || index <= targetLastIndex;
+        ++index) {
+      RenderBox child = childAfter(trailingChildWithLayout);
+      if (child == null || indexOf(child) != index) {
+        child = insertAndLayoutChild(buildChildConstraints(index),
+            after: trailingChildWithLayout);
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/flutter/extent_delegate_list.dart]: trailing comma

File: packages/devtools_app/lib/src/flutter/extent_delegate_list.dart [hunk 16]
```diff
@@ -0,0 +1,464 @@
...
+      maxPaintExtent: estimatedMaxScrollOffset,
+      // Conservative to avoid flickering away the clip during scroll.
+      hasVisualOverflow: (targetLastIndexForPaint != null &&
+              lastIndex >= targetLastIndexForPaint) ||
+          constraints.scrollOffset > 0.0,
+    );
+
+    // We may have started the layout while scrolled to the end, which would not
+    // expose a new child.
+    if (estimatedMaxScrollOffset == trailingScrollOffset)
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/flutter/extent_delegate_list.dart]: nit: add brackets to make this more readable

File: packages/devtools_app/test/flutter/extent_delegate_list_test.dart [hunk 17]
```diff
@@ -0,0 +1,238 @@
...
+
+void main() {
+  group('RenderSliverFixedExtentList', () {
+    group('extentDelegate', () {
+      test('itemExtent', () {
+        final extents = [100.0, 200.0, 50.0, 100.0];
+        final extentDelegate = FixedExtentDelegate(
+            // Create items with increasing extents.
+            computeExtent: (index) => extents[index],
+            computeLength: () => extents.length);
```
Comment (kenzieschmoll) [inline packages/devtools_app/test/flutter/extent_delegate_list_test.dart]: trailing comma here and in other tests.

File: packages/devtools_app/test/flutter/extent_delegate_list_test.dart [hunk 18]
```diff
@@ -0,0 +1,238 @@
...
+        expect(extentDelegate.minChildIndexForScrollOffset(350), 3);
+        expect(extentDelegate.minChildIndexForScrollOffset(449), 3);
+        // Off the end of the list.
+        expect(extentDelegate.minChildIndexForScrollOffset(450), 4);
+        expect(extentDelegate.minChildIndexForScrollOffset(1000000), 4);
+      });
+
+      try {
+        test('getMaxChildIndexForScrollOffset', () {
+          final extents = [100.0, 200.0, 50.0, 100.0];
```
Comment (kenzieschmoll) [inline packages/devtools_app/test/flutter/extent_delegate_list_test.dart]: put these extents in a const at the group level since they are used in several tests.

File: packages/devtools_app/test/flutter/extent_delegate_list_test.dart [hunk 19]
```diff
@@ -0,0 +1,238 @@
...
+          final extentDelegate = FixedExtentDelegate(
```
Comment (kenzieschmoll) [inline packages/devtools_app/test/flutter/extent_delegate_list_test.dart]: make this a helper: `createFixedExtentDelegate({List<double> extents = _defaultExtents})`

File: packages/devtools_app/test/flutter/extent_delegate_list_test.dart [hunk 20]
```diff
@@ -0,0 +1,238 @@
...
+              // Create items with increasing extents.
+              computeExtent: (index) => extents[index],
+              computeLength: () => extents.length);
+
+          // The behavior is a bit counter intuitive but this matching the
```
Comment (kenzieschmoll) [inline packages/devtools_app/test/flutter/extent_delegate_list_test.dart]: nit: move this comment to the case it applies to (looks like it is the 4th expect statement and the second to last?)

File: packages/devtools_app/lib/src/flutter/extent_delegate_list.dart (hunk 6 above)
Comment (jacob314) [reply packages/devtools_app/lib/src/flutter/extent_delegate_list.dart]: done

File: packages/devtools_app/lib/src/flutter/extent_delegate_list.dart (hunk 7 above)
Comment (jacob314) [reply packages/devtools_app/lib/src/flutter/extent_delegate_list.dart]: done

File: packages/devtools_app/lib/src/flutter/extent_delegate_list.dart (hunk 8 above)
Comment (jacob314) [reply packages/devtools_app/lib/src/flutter/extent_delegate_list.dart]: done

File: packages/devtools_app/lib/src/flutter/extent_delegate_list.dart (hunk 9 above)
Comment (jacob314) [reply packages/devtools_app/lib/src/flutter/extent_delegate_list.dart]: nope. removed the `as`.
This code was based on package:flutter code so there are more type checks than we would use.

File: packages/devtools_app/lib/src/flutter/extent_delegate_list.dart (hunk 10 above)
Comment (jacob314) [reply packages/devtools_app/lib/src/flutter/extent_delegate_list.dart]: done. Also removed the spurious trailing return.

File: packages/devtools_app/lib/src/flutter/extent_delegate_list.dart (hunk 16 above)
Comment (jacob314) [reply packages/devtools_app/lib/src/flutter/extent_delegate_list.dart]: done

File: packages/devtools_app/test/flutter/extent_delegate_list_test.dart (hunk 18 above)
Comment (jacob314) [reply packages/devtools_app/test/flutter/extent_delegate_list_test.dart]: if this was app code instead of test code than I would agree.
For test code, I think it matters more to keep all data close together and err more on the side of duplicating code while keeping things really clear..

File: packages/devtools_app/test/flutter/extent_delegate_list_test.dart (hunk 20 above)
Comment (jacob314) [reply packages/devtools_app/test/flutter/extent_delegate_list_test.dart]: done.

File: packages/devtools_app/test/flutter/extent_delegate_list_test.dart (hunk 19 above)
Comment (jacob314) [reply packages/devtools_app/test/flutter/extent_delegate_list_test.dart]: if this was app code I would absolutely create the helper. For  test code, I think this is slightly clearer and only a little more verbose.

File: packages/devtools_app/test/flutter/extent_delegate_list_test.dart (hunk 17 above)
Comment (jacob314) [reply packages/devtools_app/test/flutter/extent_delegate_list_test.dart]: done

File: packages/devtools_app/lib/src/flutter/extent_delegate_list.dart (hunk 12 above)
Comment (jacob314) [reply packages/devtools_app/lib/src/flutter/extent_delegate_list.dart]: yes this is the same structure as the code in the framework.

File: packages/devtools_app/lib/src/flutter/extent_delegate_list.dart (hunk 11 above)
Comment (jacob314) [reply packages/devtools_app/lib/src/flutter/extent_delegate_list.dart]: done

File: packages/devtools_app/lib/src/flutter/extent_delegate_list.dart (hunk 13 above)
Comment (jacob314) [reply packages/devtools_app/lib/src/flutter/extent_delegate_list.dart]: done.

File: packages/devtools_app/lib/src/flutter/extent_delegate_list.dart (hunk 14 above)
Comment (jacob314) [reply packages/devtools_app/lib/src/flutter/extent_delegate_list.dart]: not needed. Removed. this code is from Flutter so it is  has a few too many type checks.

File: packages/devtools_app/lib/src/flutter/extent_delegate_list.dart (hunk 15 above)
Comment (jacob314) [reply packages/devtools_app/lib/src/flutter/extent_delegate_list.dart]: done

**Comment by jacob314** [comment]
Good thing I wrote tests. One of my tests caught a bug in trying to merge the seemingly harmless if statements.

File: packages/devtools_app/lib/src/flutter/extent_delegate_list.dart [hunk 21]
```diff
@@ -0,0 +1,465 @@
...
+
+  @override
+  int get length => _offsets.length - 1;
+
+  final double Function(int index) computeExtent;
+  final int Function() computeLength;
+
+  @override
+  void recompute() {
+    final length = computeLength();
```
Comment (DaveShuckerow) [inline packages/devtools_app/lib/src/flutter/extent_delegate_list.dart]: I'm not sure this will provide any benefits in the case you're looking at where you want to animate elements in and out of the tree.

This will assume a constant height to all elements in the tree.  To use this with an animating size, you'll need a way of ensuring that the size and offset of widgets being built can change, while the size of nonbuilt widgets will not.

You'll also want to watch out for 'jumps' where recomputing the length of offscreen items has resulted in the contents of the scrollview popping up to the top of the screen.

It is relatively easy to write widget tests that cover both of these cases.

Comment (jacob314) [reply packages/devtools_app/lib/src/flutter/extent_delegate_list.dart]: The point of this class is that each item is given a custom height. Do you see a place where I am assuming a fixed height for all elements? 
`itemExtent(index)` and related methods are used to vend custom extents for each item.
With regard to animating, I have a test where the delegate animates the sizes where it is verified that the new layout matches what is expected. I can add some additional tests if you are concerned those tests aren't sufficient. Note that I already have more tests than the existing fixed extent list code this is ported from had in package:flutter. 
The ExtentDelegate is used to vend the size for the list so I expect there will be fewer edge cases where the size jumps than using a regular ListView where the total size is not known. As far as recomputing the total size, the intent is that as in the case of a ListView with a fixed extent, there isn't realy estimation of size of on-stage vs off-stage widgets. In both cases the delegate is used to determine offsets.
